### PR TITLE
util: Diagnose Lib Version #1

### LINKED ISSUE #2477: Create RPC Command to Run Diagnostics
# Feature Request

**Describe the Feature Request**
Currently, the diagnostics mainly only exists in the GUI. While the status of some of the individual components of the diagnostics can be seen in different commands, it's more work to find them and some don't exist in other commands. 

**Describe Preferred Solution**
Adding an RPC command that gives the results of the same diagnostics that exist for the GUI

### PR DESCRIPTION
**Pull request by MinaFarhan**
Closes #2477.

### CODE REVIEW

**Comment by MinaFarhan** [comment]
Create a library for Diagnose so it can be called by the RPC.

**Comment by Pythonix** [comment]
Looks like a great addition! I will pay a bounty of 1000 GRC to @MinaFarhan once this gets merged =)
Once the RPC commands are added it probably closes https://github.com/gridcoin-community/Gridcoin-Research/issues/2477?

**Comment by iFoggz** [comment]
Great to have a RPC of diagnostics, 

there is some coding changes that should be done to comply with: [coding.txt](https://github.com/gridcoin-community/Gridcoin-Research/blob/development/doc/coding.txt) as per the [CONTRIBUTING.md](https://github.com/gridcoin-community/Gridcoin-Research/blob/development/CONTRIBUTING.md). 

Once it's cleaned up I'll have a gander through as well as the other devs.

Look forward to this feature to be added.

**Comment by iFoggz** [comment]
Also should note that I'd like to see RPC diagnostics available for daemon as well.

**Comment by jamescowens** [comment]
@iFoggz He has been looking to volunteer and asked me a while back what would be a good way to get started. I suggested a smaller scale task at first, but one which would introduce him to parts of the code and not be too difficult. We settled on a split-out of the diagnostics into a service API as his first contribution. He looks to build the rpc calls as the next step.

File: src/Makefile.qt.include
Comment (jamescowens) [inline src/Makefile.qt.include]: Please use small case.

File: src/Makefile.qt.include
Comment (jamescowens) [inline src/Makefile.qt.include]: Same.

**Review by jamescowens — CHANGES_REQUESTED** [review]
This is a nice starting point. @iFoggz is right about some of the style elements. Please use the Bitcoin Core coding guide as our own contribution guide is a little out of date. Note that the VerifyTCPPort and VerifyClock functions are still using Qt plumbing. This will not work for the daemon, because Qt dependencies should not be used there. (I warned you about this, and this will be the hardest part, because you are going to have to use basic networking without the benefit of the Qt superstructure.)

**Comment by jamescowens** [comment]
Also, I have authorized your PR to do CI. Please review which build is failing and correct the cause.

**Comment by MinaFarhan** [comment]
> Looks like a great addition! I will pay a bounty of 1000 GRC to @MinaFarhan once this gets merged =) Once the RPC commands are added it probably closes #2477?

Thank you.

**Comment by MinaFarhan** [comment]
> This is a nice starting point. @iFoggz is right about some of the style elements. Please use the Bitcoin Core coding guide as our own contribution guide is a little out of date. Note that the VerifyTCPPort and VerifyClock functions are still using Qt plumbing. This will not work for the daemon, because Qt dependencies should not be used there. (I warned you about this, and this will be the hardest part, because you are going to have to use basic networking without the benefit of the Qt superstructure.)

I am thinking of calling the QApplication() from the library. This way even if the library is called from the daemon, we can initialte the Qapplication and call the QT functions , like QTimer(). What do you think?

**Comment by jamescowens** [comment]
I don't think it is going to work, because Qt is not linked into the daemon, and it is not a good idea to try and call Qt stuff in the core. The native core has scheduling functionality. The existing global scheduler (static CScheduler scheduler) can be used to initiate a "timeout" interval in the core instead of the QTimer for example. See schedule() or scheduleFromNow() in scheduler.h/cpp

**Comment by MinaFarhan** [comment]
Sounds Good!
I will change the QT functions with May be boost libraries. QTimer for example can be changed with boost deadline_timer.

**Comment by MinaFarhan** [comment]
1) Changed file names to use lower cases
2) Replaced the QT functions (ex QTimer) with boost library functions
3) Fixed the compilation error.
Next, I need to fix the code comments to match the Bitcoin standard.

File: src/qt/diagnose.h
Comment (barton2526) [inline src/qt/diagnose.h]: "teh" should be "the"

Comment (barton2526) [reply src/qt/diagnose.h]: Correct capitalization of "Pass"

File: src/qt/diagnose.h
Comment (barton2526) [inline src/qt/diagnose.h]: "runnin" should be "running"

File: src/qt/diagnose.h
Comment (barton2526) [inline src/qt/diagnose.h]: "retun" should be "return"

File: src/qt/diagnose.h
Comment (barton2526) [inline src/qt/diagnose.h]: "globa" should be "global"

File: src/qt/diagnose.h
Comment (barton2526) [inline src/qt/diagnose.h]: Did you mean CPID?

File: src/qt/diagnosticsdialog.cpp
Comment (barton2526) [inline src/qt/diagnosticsdialog.cpp]: "teests" should be "tests"

**Comment by MinaFarhan** [comment]
I have committed new code
1) changed the QT functions to the equivalent boost library so the library can be called outside UI from RPC
2) Cleaned the code and removed commented out functions
3) fixed typos in the code comment

File: src/qt/diagnose.h
Comment (MinaFarhan) [reply src/qt/diagnose.h]: Done

File: src/qt/diagnosticsdialog.cpp
Comment (MinaFarhan) [reply src/qt/diagnosticsdialog.cpp]: Done

**Comment by jamescowens** [comment]
@MinaFarhan please review the logs for CI and correct issues and get the code to compile cleanly on all architectures and pass the linter.

**Comment by MinaFarhan** [comment]
> @MinaFarhan please review the logs for CI and correct issues and get the code to compile cleanly on all architectures and pass the linter.

Yes, I am actually working on adding the RPC to the seam on as well. This requires some changes . Will submit all fixes together

**Comment by MinaFarhan** [comment]
I have added rpc command "walletdiagnose",
I also fixed some compiling warnings and errors.

File: src/Makefile.am
Comment (jamescowens) [inline src/Makefile.am]: Please keep spacing consistent.

File: src/Makefile.am
Comment (jamescowens) [inline src/Makefile.am]: Same here.

File: src/qt/diagnosticsdialog.cpp
Comment (jamescowens) [inline src/qt/diagnosticsdialog.cpp]: This is failing the linter. The code style requires

for (.....) {
}

There must be a space between the for keyword and the parenthesis, and also a space between the parenthesis and the opening brace.

File: src/qt/diagnosticsdialog.cpp
Comment (jamescowens) [inline src/qt/diagnosticsdialog.cpp]: Spelling... dignose?

**Review by jamescowens — CHANGES_REQUESTED** [review]
Nitpicking stuff, but you need to correct the code style stuff so the linter passes. There are also still some misspellings too. Please go through with a fine tooth comb. I may not have caught them all in the review. Once the code passes the CI and linter, then I will go through a more detailed pass.

File: src/qt/diagnosticsdialog.cpp
Comment (jamescowens) [reply src/qt/diagnosticsdialog.cpp]: Not done. Still teests... should be tests.

File: src/qt/diagnose.h
Comment (jamescowens) [reply src/qt/diagnose.h]: Still needs to be fixed. teh to the.

File: src/qt/diagnose.h
Comment (jamescowens) [reply src/qt/diagnose.h]: Not done. Looks like you may have lost a commit.

**Comment by jamescowens** [comment]
@MinaFarhan don't get discouraged by the "nitpicking" reviews. You will quickly get used to writing code that passes the spelling and code style stuff. The linter and CI is your friend in many ways in this regard. After your first PR, the CI will run automatically when you push a branch to github to allow you to check. You can also run the linter locally on your development workstation if you have all of the required packages installed. (This is a superset of the normal build packages.)

**Comment by MinaFarhan** [comment]
No problem :)
I will make the corrections and resubmit.

File: src/qt/diagnosticsdialog.cpp
Comment (MinaFarhan) [reply src/qt/diagnosticsdialog.cpp]: should be resolved in my next commit

File: src/qt/diagnosticsdialog.cpp
Comment (MinaFarhan) [reply src/qt/diagnosticsdialog.cpp]: I reran the linter and fixed the issues. But this one did not come out when running the linter.

File: src/Makefile.am
Comment (MinaFarhan) [reply src/Makefile.am]: It is vim that adding them :)
I have fixed the indentation in my next commit

**Comment by MinaFarhan** [comment]
I have pushed another change in an attempt to fix the linter and compiler error.
The compilation issue is because I compile locally on Ubuntu 22 but the workflow uses ubuntu 18 which has an older version of boost:asio. If it did not compile this time correctly, I will install a VM and install Ubuntu 18 on it to compile.

**Comment by jamescowens** [comment]
@MinaFarhan this is what is left on the linter (by clicking on details above and looking at the logs).

The subject line of commit hash 27fc1405361dd7376d052fc8c4d66419531bca3b is followed by a non-empty line. Subject lines should always be followed by a blank line.
The subject line of commit hash 7b22b083b096420f1fc2b9cc7a901b406d0870d5 is followed by a non-empty line. Subject lines should always be followed by a blank line.
The subject line of commit hash cc087da82abf291d9e3144adb244eae80c8d72d4 is followed by a non-empty line. Subject lines should always be followed by a blank line.
^---- failure generated from test/lint/lint-git-commit-check.sh
A new Boost dependency in the form of "boost/asio/ip/udp.hpp" appears to have been introduced:
src/wallet/diagnose.h:#include <boost/asio/ip/udp.hpp>

^---- failure generated from test/lint/lint-includes.sh
Success: no issues found in 4 source files
test/lint/lint-whitespace.sh: line 35: echo: write error: Broken pipe
This diff appears to have added new lines with trailing whitespace.
The following changes were suspected:

diff --git a/src/qt/diagnosticsdialog.h b/src/qt/diagnosticsdialog.h
@@ -49,12 +53,5 @@ private:
+
+    typedef std::set<std::pair<QLabel*, std::unique_ptr<DiagnoseLib::Diagnose>>> DiagnoseLabelTestPtr_set;
@@ -111,6 +113 @@ private slots:
+
^---- failure generated from test/lint/lint-whitespace.sh

**Comment by jamescowens** [comment]
You can fix the commit complaints by doing a rebase -i to a commit prior to those above, and then change the action to keep but change the commit comments.

File: src/wallet/rpcwallet.cpp
Comment (jamescowens) [inline src/wallet/rpcwallet.cpp:2294]: Can you please fix the name of this variable?

File: src/wallet/diagnose.h
Comment (jamescowens) [inline src/wallet/diagnose.h]: This should go at the top of the file.

Comment (jamescowens) [reply src/wallet/diagnose.h]: You should also add a copyright statement.

**Comment by MinaFarhan** [comment]
> 

Thank you, 
I have fixed the comments of the commits. I understand now why the linter was working on my local machine. It is because the commit's subject lines only appear after submitting the code

**Comment by MinaFarhan** [comment]
Finally the lint test has passed :)
Only error now is the use of the boost library. Not sure if that is important because boost Asio is needed for tcp network connection diagnostics. 

I have marked the submission ready for review .

File: src/wallet/diagnose.h
Comment (barton2526) [inline src/wallet/diagnose.h:804]: Should there be a note about MRC added here?

Comment (jamescowens) [reply src/wallet/diagnose.h:804]: That is a very good point. I would say yes.

**Comment by jamescowens** [comment]
@MinaFarhan rather than reviewing this again right now, I am creating some commits to address some things. I will do this as a PR against your branch, which then you can merge to your branch. They will flow automatically here when you do that.

**Comment by jamescowens** [comment]
@MinaFarhan I did a PR to your branch. Please take a look at the commits. Note that while it passes compile now, There are some problems in operation that you need to troubleshoot related to the boost:asio.

File: src/wallet/diagnose.h
Comment (jamescowens) [reply src/wallet/diagnose.h]: I fixed this in my PR to your branch.

**Comment by jamescowens** [comment]
@MinaFarhan after you merge my PR to your branch, you should rebase your branch on top of the head of development before pushing back here, because the root of your PR is behind development a bit at this point. Thanks.

File: src/wallet/diagnose.h
Comment (MinaFarhan) [reply src/wallet/diagnose.h:804]: Can you give me more info about MRC ? What MRC stands for and what is the proposed changes?

**Comment by jamescowens** [comment]
@MinaFarhan you should really do a rebase on top of development with your branch rather than a merge, because a merge will generate an extra merge commit that I have to filter out later in the change log.

Something like
git checkout DiagnoseLib
git reset --hard caaf95263df62c6b97ce4fa7861e6e0540f3ac46 (This is the merge of my PR to your branch.)
git checkout development
git pull
git checkout DiagnoseLib
git rebase development
git push

In general the rule is that all branches that are the source of PR's (which means the PR's themselves) should always be rebased to "advance" them to include any new commits on the base branch, while PRs should always be merged by the maintainer onto the target branch.

File: src/wallet/diagnose.h
Comment (jamescowens) [reply src/wallet/diagnose.h:804]: I resolved it in my PR to yours. :) Do a search of the main repo PRs for MRC. You will find a lot. There is also good info at gridcoin.us. MRC stands for Manual Rewards Claim.

**Comment by jamescowens** [comment]
We still need to troubleshoot the crashes on that bus error in boost asio.

**Comment by MinaFarhan** [comment]
> We still need to troubleshoot the crashes on that bus error in boost asio.

Yes, working on it.

**Comment by MinaFarhan** [comment]
> @MinaFarhan you should really do a rebase on top of development with your branch rather than a merge, because a merge will generate an extra merge commit that I have to filter out later in the change log.
> 
> Something like git checkout DiagnoseLib git reset --hard [caaf952](https://github.com/gridcoin-community/Gridcoin-Research/commit/caaf95263df62c6b97ce4fa7861e6e0540f3ac46) (This is the merge of my PR to your branch.) git checkout development git pull git checkout DiagnoseLib git rebase development git push
> 
> In general the rule is that all branches that are the source of PR's (which means the PR's themselves) should always be rebased to "advance" them to include any new commits on the base branch, while PRs should always be merged by the maintainer onto the target branch.

Ohh, sure I can do that next time. Do I still need to rebase instead of merge or is it too late?

**Comment by MinaFarhan** [comment]
> > We still need to troubleshoot the crashes on that bus error in boost asio.
> 
> Yes, working on it.

Hello
L
I can not reproduce the issue on my PC. What version of Linux you are using? And is it a random issue or always appear when diagnostics are run? 
thank you

**Comment by jamescowens** [comment]
Always. openSuSE Leap 15.4, which has boost 1.66.0.

**Comment by jamescowens** [comment]
Note that from my particular public IP, the default time service we use does not respond in a timely manner. You should simulate a non-responsive URL for the time server and see if you can reproduce the crash that way. That is the likely culprit.

**Comment by MinaFarhan** [comment]
I was not able to reproduce the issue. I have tried different situations. I turned off the internet before the test and while doing the test.
Anyway, I have added error catching for the boost asio. Can you please try again and check if it still crashes from your end?
thanks

**Comment by jamescowens** [comment]
@MinaFarhan I have some more fixes for your branch (PR) here that solves the issue I was having and some other stuff, like the Bionic problem above. The error catching that you put in does not work in Bionic, because boost 1.65 doesn't have results_type. I made some changes and I think everything works now.

**Comment by jamescowens** [comment]
@MinaFarhan I put up another PR against your branch to fix issues that were introduced when you put in the error handling.

**Comment by jamescowens** [comment]
@MinaFarhan please merge my PR and let it flow here. Then this should be ready to merge.

**Review by jamescowens — APPROVED** [review]
tACK.

**Comment by jamescowens** [comment]
This is ready to merge. The next thing on this is an rpc command to do diagnostics using this core library.

**Comment by MinaFarhan** [comment]
> I have added rpc command "walletdiagnose",

Thank you,
the RPC command is actually there. I have added rpc command "walletdiagnose". It is submitted in ["ee0fda0e4e1b44b2eb9fc26e638faf19f6adef12"](https://github.com/gridcoin-community/Gridcoin-Research/pull/2573/commits/ee0fda0e4e1b44b2eb9fc26e638faf19f6adef12)

Thanks
Mina

**Comment by jamescowens** [comment]
There is still an issue about crashing on the assert. I will track this down. assert(m_name_to_test_map.size() < Diagnose::TestSize);

**Comment by jamescowens** [comment]
This is the problem...

2022-10-31T03:31:56Z Using data directory /home/jowens/.GridcoinResearch2/testnet

2022-10-31T03:31:56Z Gridcoin version v5.4.0.2-ae6b4bc35718-dirty (release build)
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 144
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 144 registered, test map size = 1, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: VerifyWalletIsSynced: m_test_name = 2
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 2
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 2 registered, test map size = 2, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: CheckOutboundConnectionCount: m_test_name = 1
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = -150333728
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum -150333728 registered, test map size = 3, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: CheckConnectionCount: m_test_name = 0
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 2
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 2 registered, test map size = 3, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: VerifyClock: m_test_name = 8
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 1451884496
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 1451884496 registered, test map size = 4, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: CheckClientVersion: m_test_name = 3
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 2610
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 2610 registered, test map size = 5, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: VerifyBoincPath: m_test_name = 4
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 144
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 144 registered, test map size = 5, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: VerifyCPIDValid: m_test_name = 7
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 1449514896
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 1449514896 registered, test map size = 6, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: VerifyCPIDHasRAC: m_test_name = 5
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 1442843186
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 1442843186 registered, test map size = 7, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: VerifyCPIDIsActive: m_test_name = 6
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 7471220
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 7471220 registered, test map size = 8, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: VerifyTCPPort: m_test_name = 9
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 2610
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 2610 registered, test map size = 8, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: CheckDifficulty: m_test_name = 10
2022-10-31T03:31:59Z INFO: Diagnose: m_test_name = 1449260224
2022-10-31T03:31:59Z INFO: registerTest: test->m_test_name enum 1449260224 registered, test map size = 9, Diagnose::TestSize = 12
2022-10-31T03:31:59Z INFO: CheckETTS: m_test_name = 11

The constructor for Dialog is trying to register the test before the m_test_name is set.

**Comment by jamescowens** [comment]
Ok... this is better...

2022-10-31T03:43:35Z Using data directory /home/jowens/.GridcoinResearch2/testnet

2022-10-31T03:43:35Z Gridcoin version v5.4.0.2-ae6b4bc35718-dirty (release build)
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 2
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 2 registered, test map size = 1, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 1
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 1 registered, test map size = 2, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: CheckOutboundConnectionCount: m_test_name = 1
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 0
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 0 registered, test map size = 3, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: CheckConnectionCount: m_test_name = 0
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 8
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 8 registered, test map size = 4, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: VerifyClock: m_test_name = 8
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 3
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 3 registered, test map size = 5, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: CheckClientVersion: m_test_name = 3
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 4
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 4 registered, test map size = 6, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: VerifyBoincPath: m_test_name = 4
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 7
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 7 registered, test map size = 7, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: VerifyCPIDValid: m_test_name = 7
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 5
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 5 registered, test map size = 8, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: VerifyCPIDHasRAC: m_test_name = 5
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 6
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 6 registered, test map size = 9, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: VerifyCPIDIsActive: m_test_name = 6
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 9
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 9 registered, test map size = 10, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: VerifyTCPPort: m_test_name = 9
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 10
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 10 registered, test map size = 11, Diagnose::TestSize = 12
2022-10-31T03:43:37Z INFO: CheckDifficulty: m_test_name = 10
2022-10-31T03:43:37Z INFO: Diagnose: m_test_name = 11
2022-10-31T03:43:37Z INFO: registerTest: test->m_test_name enum 11 registered, test map size = 12, Diagnose::TestSize = 12

**Comment by jamescowens** [comment]
I just pushed two additional commits directly to development to fix this issue.
https://github.com/gridcoin-community/Gridcoin-Research/commit/ae6b4bc35718d5551aa97095bf9960b527c25a46
https://github.com/gridcoin-community/Gridcoin-Research/commit/4be24691c9cdb9232873eb03870140914e89c690

**Comment by jamescowens** [comment]
A few more cleanup commits directly to development branch that are followons to this PR.
https://github.com/gridcoin-community/Gridcoin-Research/commit/da333ca55fbfea08f1b9225d7da692597344cac2
https://github.com/gridcoin-community/Gridcoin-Research/commit/1239ac07a0c3a7a18b85c9cd9cf46f5847420e8a
https://github.com/gridcoin-community/Gridcoin-Research/commit/1239ac07a0c3a7a18b85c9cd9cf46f5847420e8a
https://github.com/gridcoin-community/Gridcoin-Research/commit/e325567f502d60d68fd78acede61bcfdbd88cbf9

I think there is still something not quite right with the clock test, because even with more than 5 connections the clock test does not return immediately all of the time. I will do some more investigation of that.

**Pull request merged**
